user API for compass plugin

kecnry opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our code of conduct,
https://github.com/spacetelescope/jdaviz/blob/main/CODE_OF_CONDUCT.md . -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable viz component(s). -->

This pull request extends on #1401 and provides user-API access to the compass plugin.

[Plugin User API docs](https://jdaviz--1643.org.readthedocs.build/en/1643/api/jdaviz.configs.imviz.plugins.compass.compass.Compass.html#jdaviz.configs.imviz.plugins.compass.compass.Compass)

**TODO**:
- [x] add PR to changelog if before next release, otherwise add new entry (waiting until PR is accepted otherwise to avoid constant merge conflicts)

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->


### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Are two approvals required? Branch protection rule does not check for the second approval. If a second approval is not necessary, please apply the `trivial` label.
- [x] Do the proposed changes actually accomplish desired goals? Also manually run the affected example notebooks, if necessary.
- [x] Do the proposed changes follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are tests added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Are docs added/updated as required? If so, do they follow the [STScI Style Guides](https://github.com/spacetelescope/style-guides)?
- [x] Did the CI pass? If not, are the failures related?
- [x] Is a change log needed? If yes, is it added to `CHANGES.rst`?
- [x] Is a milestone set?
- [ ] After merge, any internal documentations need updating (e.g., JIRA, Innerspace)?

duytnguyendtn reviewed (approved):
This plugin is fairly simple, so I'm happy to approve. The only consideration I MIGHT have is whether we have support for "read only" properties? It might be nice to show what data layer is being visualized via `data_label` or something like that, though we don't have a way from the UI to change it from the plugin (must be from the viewer).

duytnguyendtn commented:
(It will need a changelog entry)

kecnry commented:
> The only consideration I MIGHT have is whether we have support for "read only" properties? It might be nice to show what data layer is being visualized via data_label or something like that, though we don't have a way from the UI to change it from the plugin (must be from the viewer).

I like this idea.  As long as it's implemented as a _property_ in the plugin itself, then exposing it to the user API would still remain read-only.  I'll check to see if it would be easy to implement, or if not we can always implement it down the road.

kecnry commented on jdaviz/configs/imviz/plugins/compass/compass.py:
```diff
@@ -3,12 +3,24 @@
 __all__ = ['Compass']
 
 
 @tray_registry('imviz-compass', label="Imviz Compass")
```
do we want to rename this to just "Compass"?

duytnguyendtn replied:
Hmm... Come to think about it, maybe. Especially if we're moving towards more dynamic configurations, perhaps we should be proactive and begin stripping out the config-specific labels?

kecnry replied:
I would be in favor of that.  We may not ever put this plugin in a different tool, but if we do, I can't imagine we'd have separate plugins entirely... we would want to generalize this one.  Any opposition @pllim?  Note that this affects the dictionary key to access the API _and_ the text in the plugin tray expansion menus.

pllim replied:
I have no idea if this will work as-is on Cubeviz, Mosviz, and Specviz2d image viewers. Probably not.

kecnry replied:
No, I wouldn't expect it to out-of-the-box without some generalizing.  Is that the intention of "Imviz" in the plugin name?  Is there some other way we can document that internally without it being in the user-facing title?

pllim replied:
Naming is hard... If you want to remove "Imviz", sure, go ahead...

rosteen replied:
Remove Imviz you say? 👀 😆

pllim replied:
Re: https://github.com/spacetelescope/jdaviz/pull/1643#discussion_r969998936

I think that is a good compromise. I guess if you see a plugin in Imviz GUI already, it is pretty obvious to the user.

kecnry commented:
@duytnguyendtn  - it was already implemented as a traitlet, and so I implemented a way to have "read-only" traitlets exposed through the user API wrapper (instead of needing to wrap in a different named property in the plugin code for every case we want to do this).  Let me know what you think of this solution.

duytnguyendtn commented:
Oh wait, I must have missed it. Let me double check...

kecnry commented on jdaviz/configs/imviz/plugins/compass/compass.py:
```diff
@@ -3,12 +3,25 @@
+    """
     template_file = __file__, "compass.vue"
     icon = Unicode("").tag(sync=True)
     data_label = Unicode("").tag(sync=True)
```
is `data_label` the most appropriate here (as opposed to `layer_label`, `top_layer`, etc)?  We probably will use similar patterns in other plugins, so might want to give this a little thought.

duytnguyendtn replied:
I was thinking about that too; the reason why I said "or something like that" is because I feel like `data_label` isn't exactly the right wording, but it's the name of the traitlet. Personally I don't think its THAT horrible of a name, so I'm willing to let it slide

kecnry replied:
we can rename internal traitlets at this point easily, but once we expose it to the user-API, that would likely count as API-breaking.

duytnguyendtn replied:
Borrowing the concern from #1583, are there any other plugins that use the same traitlet name that we should be sensitive to?

kecnry replied:
I don't think any other plugins currently use this, but I think there was a suggestion/request to have it in line profiles.  @pllim would know better since relying on the "top layer" logic is pretty imviz-specific.

pllim replied:
Yes, top layer is Imviz specific for now. But... anything that uses the coordinates info panel or any astrowidgets API and allow loading multiple datasets into the same viewer should support it.

pllim replied:
If you want to be even more specific, it is actually the top image layer that is 2D and not a Subset and also not set to invisible. Name that how you will. 😸 

https://github.com/spacetelescope/jdaviz/blob/78b5d3584784a3e9cf134b506935e9e79d872064/jdaviz/configs/imviz/helper.py#L415-L425

kecnry replied:
That's a good point, it always will be an image (not another type of plotting layer/marker/etc), so maybe `data_label` is the most descriptive.

pllim reviewed (commented):
This use case seems broken but maybe out of scope?

1. Have two viewers. Show A in one and B in another, so we can tell later on.
2. Run `plugin.show('popout')` to pop out the Compass plugin.
3. Use the Compass API you added here to set the viewer. When you switch, the viewer changes but not the data label that goes with it. Seems to work if it is not popped out, I think.

Otherwise, I think it works fine. Thanks!

kecnry commented:
@pllim - I think I followed your steps, but could not reproduce.  It sounds a lot like some of the popout issues we had with ipywidgets 8 though... what version do you have installed?  Does the same workflow work if showing the plugin in a cell (with `plugin.show()`)?

pllim reviewed (approved):
I have ipywidgets 7.7.2.

Anyways, if you cannot reproduce the problem, then let's get this in.

kecnry commented:
Thanks for the reviews, I rebased on main and updated the changelog - I'll squash & merge once CI passes 

> Anyways, if you cannot reproduce the problem, then let's get this in.

I think the same bug should be possible from the UI (before this PR) by clicking the popout button in the tray, so agreed with getting this in for now and revisiting the bug if it keeps happening.  Thanks!
